### ISSUE TITLE
Toolbar icon mini app

### ISSUE THREAD

**Issue by lundman** [openzfsonwindows/ZFSin#92]
We could do with a very small app that places an icon in the bar on Windows;

![screen shot 2018-11-30 at 1655 22 2](https://user-images.githubusercontent.com/637168/49276390-bd80cb00-f4c1-11e8-85d1-8a24fa5c0bd6.png)

and initially, its menus could probably just have;

```
 > Import 
 > Status
 > Export
```

Selecting Import, runs "zpool import" to produce

```
 > Import > tank
 >        > neo
```

Selecting `tank` would run `zpool import tank`

Selecting status would run `zpool list`

```
 > Import
 > Status > tank (OK)
 >        > neo (DEGRADED)
```

Potentially allow `tank` to be selected to show `zpool status tank` with all vdevs etc.

```
 > Import
 > Status > tank (OK)
 >        > neo (DEGRADED) > 
 >        >        NAME        STATE     READ WRITE CKSUM
 >        >        tank        DEGRADED     0     0     0
 >        >          mirror-0  DEGRADED     0     0     0
 >        >            c2t2d0  ONLINE       0     0     0
 >        >            c2t1d0  UNAVAIL      0     0     0  cannot open

```

Selecting export, `zpool list` and selecting `tank` to run `zpool export tank`

```
 > Import >
 > Status >
> Export > All
>        > tank
>        > neo
```

Probably enough for simple ZFS usage with that, and can include with installer. Eventually pad it out to be able to create pools, and snapshots etc.

Now, it is famous how bad I am at doing GUIs, so feel free to make it usable.

Any takers?

**Comment by fire** [comment]
Did anyone take on the task? The task looks interesting but I know nothing about these systems.

**Comment by lundman** [comment]
Not as far as I know - if you mean ZFS side, it will be easiest to just run the processes and read the text input, assuming you know Windows dev. Hooking directly into libzfs is also possible but would need some ZFS knowledge

**Comment by fire** [comment]
libzfs sounds like the best option:

From a design standpoint, I would use the three buttons below to open a small window because a pool status would expand the list too long.

I'll see if I can draw a small design in a later post.

```
 > Import 
 > Status
 > Export
```

**Comment by fire** [comment]
According to https://docs.microsoft.com/en-us/windows/desktop/uxguide/winenv-notification the notification area shouldn't be used for commands so the design would be to have a small menu that opens a gui zfs status app.

The menu notifier menu would have:

> ZFSin Control Panel
> Exit

Double clicking on the icon will launch the ZFSin control panel.

The icon should show status. One for regular status and another icon for degraded pools.

**Comment by lundman** [comment]
Sure that sounds fine. I'm not so invested in the GUI details, as long as its useful for people - so basic things can be done without CMDline :)

If you want to link against libzpool, libzfs, and the other porting libs, that is fine (and more proper I suppose) I can help with that. Somethings are easy, if they are in libzfs_core - other things in libzfs will be a bit more hassle (not been changed into API)

**Comment by fire** [comment]
Do you know the specific functions calls for:

* Import 
  * Get list of pools
* Status
  * Get list of pools
  * Get NAME STATE READ WRITE CKSUM
* Export pools
  * Get list of pools

**Comment by lundman** [comment]
Alas, all those old commands are not in libzfs_core;

You can see how the `zpool command` work, starting in 
https://github.com/openzfsonwindows/ZFSin/blob/master/ZFSin/zfs/cmd/zpool/zpool_main.c#L6727

which in the case of `status` you can do as
```
		ret = for_each_pool(argc, argv, B_TRUE, NULL,
		    status_callback, &cb);
```
and copy the `status_callback()` (just above) function to do as you need.

**Comment by fire** [comment]
Stage two would be some sort of sun burst diagram (https://datavizcatalogue.com/methods/sunburst_diagram.html)

To show the status of the pools I would use sunburst diagrams.

![Sunburst](https://datavizcatalogue.com/methods/images/anatomy/PNG/sunburst_diagram.png)

**Comment by fire** [comment]
Off topic, but is the current zfsin ready for use on a physical windows 10 desktop or am I still recommended to use a vm?

**Comment by lundman** [comment]
`import` is trickier for sure;  To list pools found;

https://github.com/openzfsonwindows/ZFSin/blob/master/ZFSin/zfs/cmd/zpool/zpool_main.c#L2421

ends up calling `show_import()` for the probe. Or `do_import()` to attempt to import.

**Comment by lundman** [comment]
I personally use VM as you can so easily rollback snapshots to before ZFS is loaded. But you don't really need to do that since you aren't working on the kernel part. So in that sense, just install zfs, then you can play with cmdline and libzfs.  There is a guide in wiki how to delete ZFSin.sys if you get boot-crash issue, but that is unlikely as nothing uses ZFS automatically yet.

**Comment by fire** [comment]
Does calling the import / export commands require an administrator prompt?

**Comment by lundman** [comment]
Yeah unfortunately it does, I have no idea how that part works in Windows.  Eventually the commands and kernel module is supposed to be aware of it (zpool allow command etc) but for now, everything assumes Administrator.

**Comment by fire** [comment]
Do you know if zpool is compilable as a dll so I can call functions from it or do I have to link the entire zpool program without the main.

**Comment by lundman** [comment]
I skipped DLL and went for static libs for porting speed. Long term goal they could probably be changed to DLLs but I have not tried that. For today, you will need to link to at least libzpool, libzfs and libzfs_core.

**Comment by fire** [comment]
The design I have is to generate a single golang executable that static links libzpool, libzfs and libzfs_core and calls one of the golang libraries that provides the Win32 apis.

It'll have to be 32 bit for compatibility? Is 64 bit better?

The size of the executable isn't a big problem as long as it's one file which you can put into the installer.

Maybe I can reuse the code so it works for Linux, but that's a later thing.

**Comment by lundman** [comment]
I've only worked with 64bit myself - I have confirmed from time to time that 32bit works but it is not my immediate focus.

**Comment by lundman** [comment]
If you want more direct access for dev talk, we hang in #openzfs-windows on freenode

**Comment by sskras** [comment]
Very interesting thread, thanks!

**Comment by lundman** [comment]
ZOL has made some moved towards API, https://github.com/zfsonlinux/zfs/commit/e89f1295d4faa88bb05a62c8dd5f781657db5955

So that will come, let me know if soooner is better, and I can push it over
